### PR TITLE
[Formerly #468] Implement RPC-Patch-V1-01

### PR DESCRIPTION
**Pull request by bdefoy**
Validates that each patch request body contains one or more properties present in the corresponding put request body, and contains only properties present in the put request body.

Formerly #468

### CODE REVIEW

File: docs/patch-properties-in-normal-resource-definition.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+ARM OpenAPI(swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Patch-V1-01
+
+## Output Message
+
+Patch request body MUST contain at least one or more properties present in the normal resource definition (PUT operation).
```
Comment (rkmanda) [inline docs/patch-properties-in-normal-resource-definition.md:17]: >normal [](http://example.com/codeflow?start=79&length=6)

remove "normal" everywhere

Comment (rkmanda) [inline docs/patch-properties-in-normal-resource-definition.md:17]: >PUT operation [](http://example.com/codeflow?start=107&length=13)

PUT request body

File: packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    return []
+  }
+
+  const error = [
+    {
+      message: "Patch request body MUST contain at least one or more properties present in the normal resource definition (PUT operation).",
+      path: ctx.path,
+    },
+  ]
+  const patchResponses = pathItem["patch"]?.responses
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:14]: >responses [](http://example.com/codeflow?start=44&length=9)

we should be checking the request body here

File: packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  const putResponses = pathItem["put"]?.responses
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:15]: >responses [](http://example.com/codeflow?start=40&length=9)

request body

File: packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  // add all the properties of patch responses to a set
+  for (const responseStatusCode of patchResponseCodes) {
+    const propertyNames = Object.keys(getProperties(patchResponses[responseStatusCode].schema))
+    for (const name of propertyNames) {
+      patchPropertyNames.add(name)
+    }
+  }
+
+  if (patchPropertyNames.size < 1) {
+    return []
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:38]: >return [] [](http://example.com/codeflow?start=4&length=9)

Patch with empty body doesnt make sense. We should throw an error for this condition

File: packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  // defined success response codes for the put operation
+  const putResponseCodes = Object.keys(putResponses).filter((statusCode) => statusCode.startsWith("2"))
+  for (const responseStatusCode of putResponseCodes) {
+    const propertyNames = Object.keys(getProperties(putResponses[responseStatusCode].schema))
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:44]: >Object.keys [](http://example.com/codeflow?start=26&length=11)

does object.keys include the full jpath for each property?

File: packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    if (propertyNames.some((name) => patchPropertyNames.has(name))) {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:45]: >if (propertyNames.some((name) => patchPropertyNames.has(name))) { [](http://example.com/codeflow?start=4&length=65)

The condition should be that - each property defined in the patch body MUST be present in the put body

Comment (rkmanda) [reply packages/rulesets/src/spectral/functions/patch-properties-in-normal-resource-definition.ts:45]: And patch must have atleast one property

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

**Review by tejaswiMinnu — APPROVED** [review]

File: docs/patch-properties-correspond-to-put-properties.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+- RPC-Patch-V1-01
+
+## Output Message
+
+A patch request body must only contain properties present in the corresponding put request body, and must contain at least one of the properties.
+
+## Description
+
+Validates that each patch request body contains one or more properties present in the corresponding put request body, and contains only properties present in the put request body.
+
```
Comment (rkmanda) [inline docs/patch-properties-correspond-to-put-properties.md:22]: Can u pl add some valid and invalid examples here?

File: docs/rules.md [hunk 8]
```diff
@@ -784,7 +786,7 @@ Please refer to [patch-sku-property.md](./patch-sku-property.md) for details.
 ### PathCharacters
 
 Path should contain only recommended characters.
-The recommended characters are 0-9, A-Z, a-z, -, ., _, ~, and :.
+The recommended characters are 0-9, A-Z, a-z, -, ., \_, ~, and :.
```
Comment (rkmanda) [inline docs/rules.md:789]: >z, -, ., \_, ~, and  [](http://example.com/codeflow?start=43&length=20)

is this related to your change?

File: docs/rules.md [hunk 9]
```diff
@@ -1196,6 +1200,12 @@ Patch may not change the name, location, or type of the resource.
 
 Please refer to [un-supported-patch-properties.md](./un-supported-patch-properties.md) for details.
 
+### PatchPropertiesCorrespondToPutProperties
+
+Validates if Patch request body contains at least one or more properties present in the normal resource definition (PUT operation).
+
+Please refer to [patch-properties-in-normal-resource-definition.md](./patch-properties-in-normal-resource-definition.md) for details.
```
Comment (rkmanda) [inline docs/rules.md:1207]: >patch-properties-in-normal-resource-definition [](http://example.com/codeflow?start=17&length=46)

needs to be updated?

File: packages/rulesets/src/spectral/az-arm.ts [hunk 10]
```diff
@@ -272,6 +273,20 @@ const ruleset: any = {
     /// ARM RPC rules for Patch patterns
     ///
 
+    // RPC Code: RPC-Patch-V1-01
+    PatchPropertiesCorrespondToPutProperties: {
+      description:
+        "Patch request body MUST contain at least one or more properties present in the normal resource definition (PUT operation).",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:279]: >normal [](http://example.com/codeflow?start=88&length=6)

remove normal

Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:279]: >Patch request body MUST contain at least one or more properties present in the normal resource definition (PUT operation). [](http://example.com/codeflow?start=9&length=122)

Patch request body MUST contain at least one or more properties present in the resource definition as also indicated in the request or response payload of a PUT operation.

File: packages/rulesets/src/spectral/az-arm.ts [hunk 11]
```diff
@@ -272,6 +273,20 @@ const ruleset: any = {
...
+      message: "{{error}}",
+      severity: "error",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:281]: > severity: "error", [](http://example.com/codeflow?start=5&length=19)

set staging: true

Comment (rkmanda) [reply packages/rulesets/src/spectral/az-arm.ts:281]: [https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-set-a-spectral-rule-to-run-only-in-staging](https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-set-a-spectral-rule-to-run-only-in-staging)

**Comment by rkmanda** [comment]
rush build and rush change to be run

**Comment by rkmanda** [comment]
[https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-prepare-for-a-pr-submission-after-you-made-changes-locally](https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-prepare-for-a-pr-submission-after-you-made-changes-locally)

---
In reply to: [1601383409](https://github.com/Azure/azure-openapi-validator/pull/490#issuecomment-1601383409) [](http://example.com/codeflow?ancestors=1601383409)

File: packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  if (patchBodyPropertiesEmpty) {
+    return [
+      {
+        message: "Patch operations body cannot be empty.",
+        path: path,
+      },
+    ]
+  }
+
+  // patch body is empty while put body nonempty
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts:42]: >// patch body is empty while put body nonempty [](http://example.com/codeflow?start=2&length=46)

This is already covered with the if condition above

File: packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  // or patch body nonempty while put body empty => error
+  if (patchBodyPropertiesEmpty != putBodyPropertiesEmpty) {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts:44]: > if (patchBodyPropertiesEmpty != putBodyPropertiesEmpty) { [](http://example.com/codeflow?start=1&length=58)

Have an explicit condition and error code to check that the patch body is non empty and put is empty

File: packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  if (pathItem === null || typeof pathItem !== "object") {
+    return []
+  }
+
+  const path = ctx.path.concat([PATCH, PARAMETERS])
+  const error = [{ message: ERROR_MESSAGE, path: path }]
+  const errors: any = []
+
+  // array of all the patch body param properties
+  const patchBodyProperties: any[] = pathItem[PATCH]?.parameters?.filter(PARAM_IN_BODY).map((param: any) => getProperties(param.schema))
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts:25]: >getProperties [](http://example.com/codeflow?start=108&length=13)

Doesnt this just return the top level properties? We possibly need to iterate thru all properties in the patch request to check if the properties specified at each level is also present in the put request body

File: packages/rulesets/src/spectral/az-arm.ts (hunk 10 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/az-arm.ts:279]: Assuming thats same as
"Patch request body must only contain properties present in the corresponding put request body, and must contain at least one property."

File: packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts (hunk 14 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts:25]: This helper function should get all the properties 

```

 if (schema.allOf && Array.isArray(schema.allOf)) {
    schema.allOf.forEach((base: any) => {
      properties = { ...getProperties(base), ...properties }
    })
  }

```

File: docs/rules.md (hunk 8 above)
Comment (AkhilaIlla) [reply docs/rules.md]: No, I havent modified them.
I ran 
```
rush regen-ruleindex

```
and I think it pulls changes from individual .md files

File: packages/rulesets/src/spectral/az-arm.ts (hunk 11 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/az-arm.ts:281]: setting stagingOnly: true is causing build errors locally/ tests fail.
Also, I dont find any rule have this flag enabled & I believe all the rules when merged to main by default run only in Staging.

Comment (bdefoy) [reply packages/rulesets/src/spectral/az-arm.ts:281]: We do have rules merged into main with `stagingOnly` true, e.g., https://github.com/Azure/azure-openapi-validator/blob/main/packages/rulesets/src/spectral/az-arm.ts#L163. What error do you get in the tests when you add `stagingOnly`?

File: docs/patch-properties-correspond-to-put-properties.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        "schema": {
+          "$ref": "#/definitions/FooRequestParams",
+        }
+      }
+    ]
+  }
+```
+
+## Bad Example
+
```
Comment (bdefoy) [inline docs/patch-properties-correspond-to-put-properties.md:62]:
```suggestion

### Patch body parameters not found in put body

In the following example, the patch request has properties that are not present in the put request (#/definitions/FooRequestParams).

```

File: docs/patch-properties-correspond-to-put-properties.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        "name": "foo1_put",
+        "in": "body",
+        "schema": {
+          "$ref": "#/definitions/BarRequestParams",
+        },
+      }
+    ]
+  }
+```
+
```
Comment (bdefoy) [inline docs/patch-properties-correspond-to-put-properties.md:94]:
```suggestion

### Patch body empty

In the following example, the patch request has no body.

```

File: docs/patch-properties-correspond-to-put-properties.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+ARM OpenAPI(swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Patch-V1-01
+
+## Output Message
+
+A patch request body must only contain properties present in the corresponding put request body, and must contain at least one property.
```
Comment (ravimeda) [inline docs/patch-properties-correspond-to-put-properties.md:17]: Minor - Make the verbs identifiable e.g., uppercase as PUT.

File: docs/patch-properties-correspond-to-put-properties.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+        "in": "body",
+        "schema": {
+          "$ref": "#/definitions/BarRequestParams",
+        },
+      }
+    ]
+  }
+```
+
+```json
```
Comment (bdefoy) [inline docs/patch-properties-correspond-to-put-properties.md:124]: This example is almost the same as the first example, so you can probably remove it.

File: packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+  // array of all the patch body properties that are not present in the put body (if any)
+  const patchBodyPropertiesNotInPutBody = _.differenceWith(patchBodyProperties, putBodyProperties, _.isEqual)
+
+  // there is at least one property present in the patch body that is not present in the the put body => error
+  if (patchBodyPropertiesNotInPutBody.length > 0) {
+    patchBodyPropertiesNotInPutBody.forEach((missingProperty) =>
+      Object.keys(missingProperty).forEach((key) => {
+        errors.push({
+          message: `${key} property in patch body is not present in the corresponding put body.` + ERROR_MESSAGE,
```
Comment (bdefoy) [inline packages/rulesets/src/spectral/functions/patch-properties-correspond-to-put-properties.ts:59]: Missing a space

```suggestion
          message: `${key} property in patch body is not present in the corresponding put body. ` + ERROR_MESSAGE,
```

File: docs/patch-properties-correspond-to-put-properties.md (hunk 18 above)
Comment (AkhilaIlla) [reply docs/patch-properties-correspond-to-put-properties.md:124]: This one is with empty PUT and non-empty PATCH

File: packages/rulesets/src/spectral/az-arm.ts (hunk 11 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/az-arm.ts:303]: Added it!
thanks

File: packages/rulesets/src/spectral/test/patch-properties-correspond-to-put-properties.test.ts [hunk 20]
```diff
@@ -0,0 +1,572 @@
...
+                $ref: "#/definitions/BarRequestParams",
+              },
+            },
+            {
+              name: "foo1_patch",
+              in: "body",
+              schema: {
+                $ref: "#/definitions/FooProps",
+              },
+            },
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/patch-properties-correspond-to-put-properties.test.ts:153]: multiple body params are not allowed. One request may only have one body

Comment (rkmanda) [reply packages/rulesets/src/spectral/test/patch-properties-correspond-to-put-properties.test.ts:153]: Pl remove this test

File: packages/rulesets/src/spectral/test/patch-properties-correspond-to-put-properties.test.ts [hunk 21]
```diff
@@ -0,0 +1,695 @@
...
+  }
+  return linter.run(myOpenApiDocument).then((results) => {
+    expect(results.length).toBe(1)
+    expect(results[0].message).toContain("Non empty patch body with an empty put body is not valid")
+    expect(results[0].path.join(".")).toBe("paths./foo.patch.parameters")
+  })
+})
+
+test("PatchPropertiesCorrespondToPutProperties should find errors when patch has additional property in deeply nested structure", () => {
+  const myOpenApiDocument = {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/patch-properties-correspond-to-put-properties.test.ts:317]: BarRequestParams doe snot have deep nesting in the example you used. When you specify 'allOf' properties being referenced get inline at the same level

**Review by rkmanda — APPROVED** [review]
:shipit:
